onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -298,6 +311,21 @@ Status CUDAExecutionProvider::OnRunEnd() {
   auto current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventRecord(current_deferred_release_event, static_cast<cudaStream_t>(GetComputeStream())));
   CUDA_RETURN_IF_ERROR(cudaStreamSynchronize(static_cast<cudaStream_t>(GetComputeStream())));
+
+  if (shrink_default_memory_allocator_on_every_run_end_) {
+    // We currently do not support shrinking other allocators this EP may hold (like CUDA_PINNED, CUDA_CPU).
+    auto default_memory_alloc = GetAllocator(info_.device_id, OrtMemTypeDefault);
+
+    // Currently, we only support arena based allocators for usage in the CUDA EP.
+    // However, we do want to ensure that with the following sanity check
+    if (default_memory_alloc->Info().alloc_type == OrtAllocatorType::OrtArenaAllocator) {
+      // NOTE: Shrink is thread-safe
+      static_cast<IArenaAllocator*>(default_memory_alloc.get())->Shrink();
+    } else {
+      LOGS_DEFAULT(ERROR) << "The CUDA default memory allocator is not an arena-based allocator and hence cannot be shrunk";
```

Comment:
Can a check for this be made much earlier so that we don't even come here?

Reply:
Ideally we can make this assessment in the CUDA EP's ctor but it is a too early to ascertain what kind of allocator the CUDA EP is since we would only know while registering the allocator. I actually think this check is not needed - we won't be able to support non-arena based allocator for the CUDA EP for now. Let me just remove this check ?